controller/src/main/java/io/pravega/controller/store/stream/PersistentStreamBase.java

```diff
@@ -291,33 +306,124 @@ public String getScopeName() {
      *
      * @param newRanges      key ranges of new segments to be created
      * @param scaleTimestamp scaling timestamp
-     * @return : list of newly created segments
+     * @param runOnlyIfStarted run only if the scale operation was started.
+     * @return : list of newly created segments with current epoch
      */
     @Override
-    public CompletableFuture<List<Segment>> startScale(final List<Integer> sealedSegments,
-                                                       final List<AbstractMap.SimpleEntry<Double, Double>> newRanges,
-                                                       final long scaleTimestamp) {
-        return verifyLegalState(getSegmentChunks()
-                .thenCompose(this::getLatestChunk)
-                .thenCompose(latestSegmentData -> addNewSegments(newRanges, scaleTimestamp, latestSegmentData))
-                .thenCompose(startingSegmentNumber -> getSegments(IntStream.range(startingSegmentNumber,
-                        startingSegmentNumber + newRanges.size())
+    public CompletableFuture<StartScaleResponse> startScale(final List<Integer> sealedSegments,
+                                                            final List<AbstractMap.SimpleEntry<Double, Double>> newRanges,
+                                                            final long scaleTimestamp,
+                                                            boolean runOnlyIfStarted) {
+        return verifyState(() -> getHistoryTable()
+                .thenCompose(historyTable -> getSegmentTable().thenApply(segmentTable -> new ImmutablePair<>(historyTable, segmentTable)))
+                .thenCompose(pair -> {
+                    final Data<T> segmentTable = pair.getRight();
+                    final Data<T> historyTable = pair.getLeft();
+                    final int activeEpoch = TableHelper.getActiveEpoch(historyTable.getData()).getKey();
+
+                    final int nextSegmentNumber;
+                    if (TableHelper.isScaleOngoing(historyTable.getData(), segmentTable.getData())) {
+                        if (TableHelper.isRerunOf(sealedSegments, newRanges, historyTable.getData(), segmentTable.getData())) {
+                            // rerun means segment table is already updated. No need to do anything
+                            nextSegmentNumber = TableHelper.getSegmentCount(segmentTable.getData()) - newRanges.size();
+                            return CompletableFuture.completedFuture(
+                                    new ImmutablePair<>(activeEpoch, nextSegmentNumber));
+                        } else {
+                            throw new ScaleOperationExceptions.ScaleStartException();
+                        }
+                    } else {
+                        // check input is valid
```

Comment:
I think this method could be simplified by creating some private methods. This would one good one.